### PR TITLE
NOISSUE - Add Fetch chart data functionality

### PR DESCRIPTION
**Pull request by Musilah**
# What type of PR is this?
This is a Feature that updates the dashboards charts to be able to get data from the backend. 

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Dependency Update

## What does this do?
It adds fetch chart data functionality to the UI. This will facilitate the fetching of data from the timescale db via the timescale reader service.

## Which issue(s) does this PR fix/relate to?

- Related Issue #
- Resolves #

## Have you included tests for your changes?

- [ ] Yes
- [X] No 

## Did you document any new/modified functionality?

- [ ] Yes
- [x] No

### CODE REVIEW

File: ui/service.go [hunk 1]
```diff
@@ -1833,6 +1830,20 @@ func (us *uiService) ReadMessages(s Session, channelID, thingKey string, page, l
...
 
+func (us *uiService) FetchChartData(token string, channelID string, mpgm sdk.MessagePageMetadata) ([]byte, error) {
+	msg, err := us.sdk.ReadMessages(mpgm, channelID, token)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	data, jsonErr := json.Marshal(msg)
+	if jsonErr != nil {
+		return []byte{}, errors.Wrap(err, ErrExecTemplate)
```
Comment (1998-felix) [inline ui/service.go:1841]: Could you wrap with a more suitable err?

**Review by 1998-felix — APPROVED** [review]

File: ui/api/logging.go [hunk 2]
```diff
@@ -1161,13 +1161,13 @@ func (lm *loggingMiddleware) Publish(channelID, thingKey string, message ui.Mess
...
 // ReadMessages adds logging middleware to read messages method.
-func (lm *loggingMiddleware) ReadMessages(s ui.Session, channelID, thingKey string, page, limit uint64) (b []byte, err error) {
+func (lm *loggingMiddleware) ReadMessages(s ui.Session, channelID, thingKey string, page uint64, mpgm sdk.MessagePageMetadata) (b []byte, err error) {
 	defer func(begin time.Time) {
 		args := []any{
 			slog.String("duration", time.Since(begin).String()),
 			slog.String("channel_id", channelID),
 			slog.Uint64("page", page),
-			slog.Uint64("limit", limit),
+			slog.Any("page metadata", mpgm),
```
Comment (rodneyosodo) [inline ui/api/logging.go:1170]:
```suggestion
			slog.Any("page_metadata", mpgm),
```

File: ui/api/errors.go [hunk 3]
```diff
@@ -6,8 +6,9 @@ package api
 import "github.com/absmach/magistrala/pkg/errors"
 
 var (
-	errAuthorization          = errors.New("missing or invalid credentials provided")
+	errCredentials            = errors.New("missing or invalid credentials provided")
```
Comment (rodneyosodo) [inline ui/api/errors.go:9]:
```suggestion
	errInvalidCredentials            = errors.New("missing or invalid credentials provided")
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/service.go [hunk 4]
```diff
@@ -371,7 +372,9 @@ type Service interface {
 	// Publish facilitates a thing publishin messages to a channel.
 	Publish(channelID, thingKey string, message Message) error
 	// ReadMessages retrieves messages published in a channel.
-	ReadMessages(s Session, channelID, thingKey string, page, limit uint64) ([]byte, error)
+	ReadMessages(s Session, channelID, thingKey string, page uint64, mpgm sdk.MessagePageMetadata) ([]byte, error)
+	// FetchChartData retrieves messages published in a channel to populate charts.
+	FetchChartData(token string, channelID string, mpgm sdk.MessagePageMetadata) ([]byte, error)
```
[the remark this answers is not in the document]
Comment (ianmuchyri) [reply ui/service.go:377]: yes, they call the same SDK function but don't return the same response. `ReadMessages` renders an `HTML` template, while `FetchChartData` renders a `JSON` object that will be used in populating the chart. No HTML is being rendered in the latter.

[the remark this answers is not in the document]
Comment (dborovcanin) [reply ui/service.go:377]: This is OK at the moment, but in the future, we should consider using JS SDK to access MG directly from our browser. That will require changes in session handling due to encrypted cookies.

**Review by dborovcanin — APPROVED** [review]

File: ui/service.go [hunk 5]
```diff
@@ -371,7 +372,9 @@ type Service interface {
 	// Publish facilitates a thing publishin messages to a channel.
 	Publish(channelID, thingKey string, message Message) error
 	// ReadMessages retrieves messages published in a channel.
-	ReadMessages(s Session, channelID, thingKey string, page, limit uint64) ([]byte, error)
+	ReadMessages(s Session, channelID, thingKey string, page uint64, mpgm sdk.MessagePageMetadata) ([]byte, error)
```
Comment (rodneyosodo) [inline ui/service.go:375]: Isn't the page contained in `MessagePageMetadata` as offset?

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/service.go [hunk 6]
```diff
@@ -1833,6 +1833,20 @@ func (us *uiService) ReadMessages(s Session, channelID, thingKey string, page, l
 	return btpl.Bytes(), nil
 }
 
+func (us *uiService) FetchChartData(token string, channelID string, mpgm sdk.MessagePageMetadata) ([]byte, error) {
+	msg, err := us.sdk.ReadMessages(mpgm, channelID, token)
```
Comment (rodneyosodo) [inline ui/service.go:1837]:
```suggestion
	msg, sdkErr := us.sdk.ReadMessages(mpgm, channelID, token)
```

The for json leave it as `err`

File: ui/api/requests.go [hunk 7]
```diff
@@ -646,27 +655,38 @@ type readMessagesReq struct {
...
+	if req.AccessToken == "" && req.thingKey == "" {
+		return errInvalidCredentials
 	}
 	if req.channelID == "" {
 		return errMissingChannelID
 	}
-	if req.thingKey == "" {
-		return errMissingThingKey
-	}
-	if req.page == 0 {
```
Comment (rodneyosodo) [inline ui/api/requests.go:663]: No need to remove this check just use a different variable

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/api/metrics.go [hunk 8]
```diff
@@ -670,13 +670,23 @@ func (mm *metricsMiddleware) Publish(channelID, thingKey string, message ui.Mess
...
 	}(time.Now())
 
-	return mm.svc.ReadMessages(s, channelID, thingKey, page, limit)
+	return mm.svc.ReadMessages(s, channelID, thingKey, mpgm)
+}
+
+// FetchChartData adds metrics middleware to fetch chart data method.
+func (mm *metricsMiddleware) FetchChartData(token string, channelID string, mpgm sdk.MessagePageMetadata) ([]byte, error) {
+	defer func(begin time.Time) {
+		mm.counter.With("method", "fetch_reader_data").Add(1)
```
Comment (rodneyosodo) [inline ui/api/metrics.go:685]:
```suggestion
		mm.counter.With("method", "fetch_chart_data").Add(1)
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — APPROVED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**